File: RubyCivitas/pkg/RubyApplication1-0.0.1/lib/mazo_sorpresas.rb
require_relative "diario"
require_relative "sorpresa"

module Civitas
  class MazoSorpresas

    def init

      @sorpresas.clear
      @cartas_especiales.clear
      @barajada = false
      @usadas = 0

    end

    def initialize(d = false)
      @sorpresas = Array.new { Sorpresa }
      @barajada = false
      @usadas = nil
      @debug = d
      @cartas_especiales = Array.new { Sorpresa }
      @ultima_sorpresa = nil
      init
    end

    def al_mazo(s)

      unless @barajada

        @sorpresas.push(s)

      end

    end

    def siguiente

      if !@barajada || @usadas == @sorpresas.size && !@debug

        @usadas = 0
        @barajada = true

      end

      @usadas += 1
      first = @sorpresas.at(0)
      @sorpresas.delete_at(0)
      @sorpresas.push(first)
      @ultima_sorpresa = first

      @sorpresas.shuffle

      first

    end

    def inhabilitar_carta_especial(s)

      if @sorpresas.include?(s)

        card = @sorpresas.at(@sorpresas.find_index(s))
        @sorpresas.delete(s)
        @cartas_especiales.push(card)

        Diario.instance.ocurre_evento("[DIARIO] Inhabilitacion carta especial")

      end

    end

    def habilitar_carta_especial(s)

      if @cartas_especiales.include?(s)

        card = @cartas_especiales.at(@cartas_especiales.find_index(s))
        @cartas_especiales.delete(s)
        @sorpresas.push(card)

        Diario.instance.ocurre_evento("[DIARIO] Habilitacion carta especial")

      end

    end
  end
end

File: JavaCivitas/src/civitas/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;
import java.util.ArrayList;

/**
 *
 * @author emilio
 */
public class MazoSorpresas {
    
    private ArrayList<Sorpresa> sorpresas = new ArrayList<Sorpresa>();
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales = new ArrayList<Sorpresa>();
    private Sorpresa ultimaSorpresa;
    
    
    
    void init(){
        sorpresas.clear();
        cartasEspeciales.clear();
        barajada = false;
        usadas = 0;
    }
    
    MazoSorpresas(boolean debugger){
        debug = debugger;
        init();
        if(debugger){
            Diario.getInstance().ocurreEvento("\n[DIARIO] Modo DEBUG activado");
        }
        
    }
    
    MazoSorpresas(){
        init();
        debug=false;
    }
    
    
    //metodos visibilidad de paquete
    
    void alMazo(Sorpresa s){
        if(!barajada){
            sorpresas.add(s);
        }
    }
    
    Sorpresa siguiente(){
        if(!barajada || usadas == sorpresas.size()){
            if(!debug){
                barajada=true;
                //falta poner el metodo para barajar
                usadas=0;
            }
            usadas++;
            Sorpresa s = sorpresas.get(0);
            sorpresas.add(s);
            ultimaSorpresa = s;
            sorpresas.remove(0);
            
        }
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        if(cartasEspeciales.contains(sorpresa)){
            sorpresas.add(sorpresa);
            cartasEspeciales.remove(sorpresa);
            Diario.getInstance().ocurreEvento("\n[DIARIO]Carta especial inhabilitada.");
        }
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        if(cartasEspeciales.contains(sorpresa)){
            sorpresas.add(sorpresa);
            cartasEspeciales.remove(sorpresa);
            Diario.getInstance().ocurreEvento("\n[DIARIO] Carta especial habilitada.");
            
        }
    }
    
    Sorpresa getUltimaSorpresa(){
        return sorpresas.get(sorpresas.size()-1);
    }
    
}
